python/ribasim/tests/test_io.py

```diff
@@ -18,28 +23,61 @@ def assert_equal(a, b):
 
 
 def test_basic(basic, tmp_path):
-    model = basic
-    model.write(tmp_path / "basic")
-    model = ribasim.Model.from_toml(tmp_path / "basic/basic.toml")
+    model_orig = basic
+    model_orig.write(tmp_path / "basic")
+    model_loaded = ribasim.Model.from_toml(tmp_path / "basic/basic.toml")
 
-    assert model.modelname == "basic"
-    index_a = basic.node.static.index.to_numpy(int)
-    index_b = model.node.static.index.to_numpy(int)
+    assert model_orig.modelname == model_loaded.modelname
+    index_a = model_orig.node.static.index.to_numpy(int)
+    index_b = model_loaded.node.static.index.to_numpy(int)
     assert_array_equal(index_a, index_b)
-    assert_equal(basic.node.static, model.node.static)
-    assert_equal(basic.edge.static, model.edge.static)
-    assert model.basin.forcing is None
+    assert_equal(model_orig.node.static, model_loaded.node.static)
+    assert_equal(model_orig.edge.static, model_loaded.edge.static)
+    assert model_loaded.basin.forcing is None
 
 
 def test_basic_transient(basic_transient, tmp_path):
-    model = basic_transient
-    model.write(tmp_path / "basic-transient")
-    model = ribasim.Model.from_toml(tmp_path / "basic-transient/basic-transient.toml")
+    model_orig = basic_transient
+    model_orig.write(tmp_path / "basic-transient")
+    model_loaded = ribasim.Model.from_toml(
+        tmp_path / "basic-transient/basic-transient.toml"
+    )
 
-    assert model.modelname == "basic-transient"
-    assert_equal(basic_transient.node.static, model.node.static)
-    assert_equal(basic_transient.edge.static, model.edge.static)
+    assert model_orig.modelname == model_loaded.modelname
+    assert_equal(model_orig.node.static, model_loaded.node.static)
+    assert_equal(model_orig.edge.static, model_loaded.edge.static)
 
-    forcing = model.basin.forcing
-    assert_equal(basic_transient.basin.forcing, forcing)
+    forcing = model_loaded.basin.forcing
+    assert_equal(model_orig.basin.forcing, forcing)
     assert forcing.shape == (1468, 8)
+
+
+def test_input_mixin():
+    class Schema(pa.SchemaModel):
+        col_x: Series[int] = pa.Field(coerce=True)
+        col_y: Series[float]
+
+    class Dummy(ribasim.input_base.InputMixin, BaseModel):
+        myfield_a: int
+        myfield_b: float
+        myfield_c: DataFrame[Schema]
+
+    proper_c = pd.DataFrame(data={"col_x": [1], "col_y": [1.0]})
+    bad_c = pd.DataFrame(data={"col_y": [1.0]})
+
+    dummy_1 = Dummy(myfield_a=1, myfield_b=1, myfield_c=proper_c)
+
+    # Type cast if possible is allowed (myfield_b int -> float)
+    assert (
+        repr(dummy_1)
+        == "<ribasim.Dummy>\n   myfield_a: 1\n   myfield_b: 1.0\n   myfield_c: DataFrame(rows=1) (col_x, col_y)"
+    )
+
+    with pytest.raises(ValidationError) as exc_info:
+        Dummy(myfield_a=1, myfield_b=1, myfield_c=bad_c)
+
+    assert exc_info.value.errors()[0]["loc"] == ("myfield_c",)
```

Comment:
My gut feeling is that this test is a bit too low level.

Instead of creating a new class that implements `InputMixin`, let's use an existing class and check it's `repr`.